tests/conftest.py

```diff
@@ -89,3 +94,54 @@ def _snapshot(**kwargs):
             yield snapshot
 
     return _snapshot
+
+
+def run_function_from_file(item):
+    file, _, func = item.location
+    marker = item.get_closest_marker("run")
+    expected_status = marker.kwargs.get("status", 0)
+    expected_out = marker.kwargs.get("out", "").encode("utf-8")
+    expected_err = marker.kwargs.get("err", "").encode("utf-8")
```

Comment:
I wonder if we should also offer an assertion function to be called with these as input.

I know we have some that need to do regex against `out`/`err`.


Or.... we could yield the function to be called:

```python
@pytest.markers.run
def test_prog_register():
    def _():
        from ddtrace.internal import atexit

        def foobar(what):
            print("hello")
            print(what)

        atexit.register(foobar, "world")

    status, out, err, pid = yield _
```

I actually don't think I love this, just a random idea.

Reply:
I was thinking we could pass an arbitrary expression as a literal string if we really need to do something special with `out`, `err`, `status` or `pid`. But perhaps we could just assume that we are passing a regex.

Reply:
regex would probably be ok.

and we can always have it where the existing code is still available if you need to do anything special.

Reply:
we could also allow for lambdas for slightly more generic expressions

Reply:
you know... I think maybe just refactor so we have a generic helper to execute a Python function in a subprocess.

```python
def test_prog_register(run_func_in_subprocess):
    def func_to_run():
        from ddtrace.internal import atexit

        def foobar(what):
            print("hello")
            print(what)

        atexit.register(foobar, "world")

    status, out, err, pid = run_func_in_subprocess(func_to_run, **kwargs)
```

and then the `@pytest.mark.run` marker will just do that for the whole function like you have it. then we can be restrictive, like `out`/`err` are regex, and `status` is a specific number. if you need to do anything else fancy, you can use this helper to execute the function and get the results yourself.

how do you feel about that?

Reply:
> how do you feel about that?

Looking at the existing tests, there doesn't even seem to be the need of regexs. It looks like, in principle, every test can be written so that one could just check the status code, as most of the testing logic could be pushed to the subprocess with asserts. Hence I'd be inclined to maybe leave the helper around, but write the tests to avoid it. WDYT?

Reply:
oh yeah, I'd rather go for the marker which should be easiest.

I am just thinking about cases where we want to test logging output where we do need out/err, but otherwise agree most can probably do fine with just status checking.

I am thinking about the `ddtrace-run` tests which do `print("SUCCESS")` and then we assert that `out == "SUCCESS"` instead of using the script status code.